jax/_src/lax/linalg.py

```diff
@@ -381,13 +408,28 @@ def eig_batching_rule(batched_args, batch_dims, *, compute_left_eigenvectors,
                      compute_right_eigenvectors=compute_right_eigenvectors),
           (0,) * (1 + compute_left_eigenvectors + compute_right_eigenvectors))
 
+def eig_jvp_rule(primals, tangents, *, compute_left_eigenvectors,
+                 compute_right_eigenvectors):
+  if compute_left_eigenvectors or compute_right_eigenvectors:
+    raise NotImplementedError(
+        'The derivatives of eigenvectors are not implemented, only '
+        'eigenvalues. See '
+        'https://github.com/google/jax/issues/2748 for discussion.')
+  # Formula for derivative of eigenvalues w.r.t. a is eqn 4.60 in
+  # https://arxiv.org/abs/1701.00392
+  a, = primals
+  da, = tangents
+  l, v = eig(a, compute_left_eigenvectors=False)
+  return [l], [jnp.sum(_solve(v, da.astype(v.dtype)) * _T(v), -1)]
```

Comment:
Rather than multiply followed by sum, I wonder if we should use a batched vector dot-product here instead? i.e.,
`einsum('ij,ji->i', _solve(v, da.astype(v.dtype), v)`

(or the equivalent using the lower level `lax.dot_general`)

This gives slightly higher level information to XLA, which in principle provides more opportunities for optimization. That said, when I test it out performance seems to be slightly worse! I guess we should leave it alone, then :)

e.g., on CPU:
```python
import jax.numpy as jnp
import jax
import numpy as np

N = 512
x = jnp.array(np.ones((N, N)))
y = jnp.array(np.ones((N, N)))

@jax.jit
def f(x, y):
  return jnp.einsum('ij,ji->i', x, y)

@jax.jit
def g(x, y):
  return (x * y.T).sum(axis=-1)

%timeit f(x, y).block_until_ready()  # 1000 loops, best of 3: 969 µs per loop
%timeit g(x, y).block_until_ready()  # 1000 loops, best of 3: 660 µs per loop
```

Reply:
This is the kind of thing that I imagine XLA can fuse, I'm slightly surprised that its worse with the user-fused code though. Maybe the jaxpr can tell us something...
```python
In [1]: import jax.numpy as jnp
   ...: import jax
   ...: import numpy as np
   ...:
   ...: N = 512
   ...: x = jnp.array(np.ones((N, N)))
   ...: y = jnp.array(np.ones((N, N)))
   ...:
   ...: @jax.make_jaxpr
   ...: def f(x, y):
   ...:   return jnp.einsum('ij,ji->i', x, y)
   ...:
   ...: @jax.make_jaxpr
   ...: def g(x, y):
   ...:   return (x * y.T).sum(axis=-1)
   ...:
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
In [2]: f(x, y)
Out[2]:
{ lambda  ; a b.
  let c = xla_call[ backend=None
                    call_jaxpr={ lambda  ; a b.
                                 let c = dot_general[ dimension_numbers=(((0,), (1,)), ((1,), (0,)))
                                                      precision=None ] b a
                                 in (c,) }
                    device=None
                    donated_invars=(False, False)
                    name=_einsum ] a b
  in (c,) }
In [3]: g(x, y)
Out[3]:
{ lambda  ; a b.
  let c = transpose[ permutation=(1, 0) ] b
      d = mul a c
      e = reduce_sum[ axes=(1,) ] d
  in (e,) }
```
So it seems dot_general on CPU lowers to something slower than the alternative. I would guess this isn't the case on other backends, but I'm gonna leave it as it is for now anyway.